core/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java

```diff
@@ -78,21 +79,49 @@ public NodeConnectionsService(Settings settings, ThreadPool threadPool, Transpor
         this.reconnectInterval = NodeConnectionsService.CLUSTER_NODE_RECONNECT_INTERVAL_SETTING.get(settings);
     }
 
-    public void connectToNodes(Iterable<DiscoveryNode> discoveryNodes) {
-
-        // TODO: do this in parallel (and wait)
+    public void connectToNodes(DiscoveryNodes discoveryNodes) {
+        CountDownLatch latch = new CountDownLatch(discoveryNodes.getSize());
         for (final DiscoveryNode node : discoveryNodes) {
+            final boolean connected;
             try (Releasable ignored = nodeLocks.acquire(node)) {
                 nodes.putIfAbsent(node, 0);
-                validateNodeConnected(node);
+                connected = transportService.nodeConnected(node);
```

Comment:
yeah, that's a bit of a pain. The handshaking introduces blocking connections and we protect against doing that on the cluster state thread. Instead of whitelisting it, I opted to go for a long standing TODO and setup the connections concurrently (I saw it help in the 100 nodes cluster tests we made long ago). However, I didn't want to spawn threads on every cluster state update just to check that nodes are still connected and up to date.